tracepath.c

```diff
@@ -124,6 +130,85 @@ static void print_host(struct run_state const *const ctl, char const *const a,
 	printf("%*s", HOST_COLUMN_SIZE - plen, "");
 }
 
+#ifdef SO_EE_RFC4884_FLAG_INVALID
+/*
+ * Convert an RFC5837 "role" field to a printable string.
+ */
+static char const *print_role(const uint8_t role)
+{
+	switch (role) {
+	case 0:
+		return _("Arrival interface");
+	case 1:
+		return _("Arrival interface sub-IP component");
+	case 2:
+		return _("Forwarding interface");
+	case 3:
+		return _("IP next hop");
+	default:
+		return _("Unkown role");
+	}
+}
+
+static void print_interface_identification(char *pkt, unsigned pkt_len, unsigned offset)
+{
+	char if_addr[INET_ADDRSTRLEN] = "";
+	uint8_t if_index_flag, ipaddr_flag, name_flag, mtu_flag, name_len, role;
+	uint32_t if_index, mtu;
+
+	offset += 4;
+	while (offset + sizeof(struct icmp_extobj_hdr) <= pkt_len) {
+		struct icmp_extobj_hdr *obj_hdr = (struct icmp_extobj_hdr *) &pkt[offset];
+		offset += sizeof(struct icmp_extobj_hdr);
+		if (obj_hdr->class_num == 2) {
+			/* This is an Interface Identification object.*/
+			role = obj_hdr->class_type >> 6;
+			if_index_flag = (obj_hdr->class_type >> 3) & 1;
+			ipaddr_flag = (obj_hdr->class_type >> 2) & 1;
+			name_flag = (obj_hdr->class_type >> 1) & 1;
+			mtu_flag = (obj_hdr->class_type >> 0) & 1;
+			printf("\t%s:", print_role(role));
+
+			if (if_index_flag) {
+				if (offset + sizeof(uint32_t) > pkt_len)
+					goto err;
+				if_index = ntohl(*((uint32_t *) &pkt[offset]));
+				printf(" index=%d;", if_index);
+				offset += 4;
+			}
+			if (ipaddr_flag) {
+				if (offset + 8 > pkt_len || ntohs(*(uint16_t *)(pkt + offset)) != 1)
+					goto err;
+				printf(" address=%s;", inet_ntop(AF_INET, &pkt[offset + 4], if_addr, sizeof(if_addr)));
```

Comment:
Only called `setsockopt` with `IP_RECVERR_RFC4884` not the IPv6 equivalent, so any interface returned in a well-formed packet will be IPv4.

Why not `setsockopt` with `IPV6_RECVERR_RFC4884`? I haven't had time to test it yet. It will be a relatively painless patch once I can test it.